app/src/main/java/tech/relaycorp/gateway/App.kt

```diff
@@ -117,9 +121,11 @@ open class App : Application() {
         Security.insertProviderAt(Conscrypt.newProvider(), 1)
     }
 
-    private fun registerGateway() {
-        if (mode == Mode.Test) return
+    private fun bootstrapGateway() {
         ioScope.launch {
+            localConfig.generateKeyPair()
+            localConfig.generateCargoDeliveryAuth()
```

Comment:
I originally implemented this in a new method inside `App`, but I was running into what appears to be a race condition because `registerGateway` and my new method were executed in different threads. This caused the key pair to be generated twice, often causing the cargo delivery auth certificate to be issued with the old key pair.

So I'm repurposing this method so that I can use the same scope.

Reply:
Right now this is being called on every app start, and not only as needed. So you're going to get new keypairs every time. The methods should check themselves if they are needed, or we should check before calling them.

Also, they won't run in instrumented tests, which I'm not sure that's what we want. We don't want to run the registration, because that requires other apps in the phone, but generating the keypair I think we want to.

Reply:
> Right now this is being called on every app start, and not only as needed. So you're going to get new keypairs every time. The methods should check themselves if they are needed, or we should check before calling them.

Good catch! We certainly don't want that.

> Also, they won't run in instrumented tests, which I'm not sure that's what we want.  We don't want to run the registration, because that requires other apps in the phone, but generating the keypair I think we want to.

Good point. Maybe the "gateway bootstrap" should be a suspend function like:

```
suspend fun bootstrapGateway() {
    localConfig.generateKeyPair()
    localConfig.generateCargoDeliveryAuth()
    if (mode != Mode.Test) {
        registerGateway.registerIfNeeded()
    }
}
```

And it could then be called from within a `ioScope.launch {}` block in `App.kt`, and also instrumentation tests. And this would be done in such a way that the keys are generated before attempting to register with the public gateway.

Thoughts?

Reply:
The `suspend` makes sense. You might want to add the `@VisibleForTesting` annotation, because it should be a private method, but we want to use it for tests too.

But do note that, although we only want to generate a keypair once, we want to call registerIfNeeded at every app start.